Add MongoDB Atlas blueprint example (#2013)
kanisterio/kanister: examples/mongodb-atlas/README.md

database and data services to accelerate and simplify how you build with data.
It deploys and scales a MongoDB cluster in the cloud. MongoDB Atlas can be
primarily classified under "MongoDB Hosting".

## Prerequisites

* Kubernetes 1.20+
* Kanister controller version 0.90.0 installed in your cluster
* Kanctl CLI installed (https://docs.kanister.io/tooling.html#install-the-tools)
* MongoDB Atlas account registered with organization, project and cluster

## Integrating with Kanister

In case, if you don't have `Kanister` installed already, use following commands
to install.
```bash
$ helm repo add kanister https://charts.kanister.io
$ helm install kanister --namespace kanister --create-namespace \
    kanister/kanister-operator --set image.tag=0.90.0
```

### Build docker image for MongoDB Atlas
To build docker image for MongoDB Atlas, execute following steps.
```bash
$ cd ~/kanister/docker/mongodb-atlas
$ docker build -t <registry_name>/<account_name>/mongodb-atlas:<tag_name> .
$ docker push <registry_name>/<account_name>/mongodb-atlas:<tag_name>
```

### Create Blueprint
Create Blueprint in the same namespace as the controller

> **NOTE:**

Author: It's a different section altogether. Do we need a note for this?

> **NOTE:**

Author: Are we saying "put a note before build image section" or "put a note to suggest build image first before creating BP"?

# MongoDB Atlas

[MongoDB Atlas](https://www.mongodb.com/atlas) is an integrated suite of cloud
database and data services to accelerate and simplify how you build with data.
It deploys and scales a MongoDB cluster in the cloud. MongoDB Atlas can be
primarily classified under "MongoDB Hosting".

## Prerequisites

* Kubernetes 1.20+
* Kanister controller version 0.90.0 installed in your cluster
* Kanctl CLI installed (https://docs.kanister.io/tooling.html#install-the-tools)
* MongoDB Atlas account registered with organization, project and cluster

## Integrating with Kanister

In case, if you don't have `Kanister` installed already, use following commands
to install.
```bash
$ helm repo add kanister https://charts.kanister.io
$ helm install kanister --namespace kanister --create-namespace \
    kanister/kanister-operator --set image.tag=0.90.0
```

> ### Build docker image for MongoDB Atlas
> To build docker image for MongoDB Atlas, execute following steps.
> ```bash
> $ cd ~/kanister/docker/mongodb-atlas
> $ docker build -t <registry>/<account_name>/mongodb-atlas:<tag_name> .
> $ docker push <registry>/<account_name>/mongodb-atlas:<tag_name>
> ```

Reviewer: ```suggestion
### Build docker image for MongoDB Atlas

To build docker image for MongoDB Atlas, execute following steps.
```bash
$ cd ~/kanister/docker/mongodb-atlas
$ docker build -t <registry>/<account_name>/mongodb-atlas:<tag_name> .
$ docker push <registry>/<account_name>/mongodb-atlas:<tag_name>
```
```

Author: updated `build image` section

# MongoDB Atlas

> [MongoDB Atlas](https://www.mongodb.com/atlas) is an integrated suite of cloud
> database and data services to accelerate and simplify how you build with data.
> It deploys and scales a MongoDB cluster in the cloud. MongoDB Atlas can be
> primarily classified under "MongoDB Hosting".

Reviewer: ```suggestion
[MongoDB Atlas](https://www.mongodb.com/atlas) is an integrated suite of cloud
database and data services to accelerate and simplify how you build with data.
It deploys and scales a MongoDB cluster in the cloud.
```

Author: fixed

## Prerequisites

* Kubernetes 1.20+
* Kanister controller version 0.90.0 installed in your cluster
* Kanctl CLI installed (https://docs.kanister.io/tooling.html#install-the-tools)
* MongoDB Atlas account registered with organization, project and cluster

## Integrating with Kanister

In case, if you don't have `Kanister` installed already, use following commands
to install.
```bash
$ helm repo add kanister https://charts.kanister.io
$ helm install kanister --namespace kanister --create-namespace \
    kanister/kanister-operator --set image.tag=0.90.0
```

> ### Build docker image for MongoDB Atlas
> 
> To build docker image for MongoDB Atlas, execute following steps. Image build is
> needed as connection process to MongoDB Atlas is using `mongosh` command. Also,
> it installs `atlas` CLI which is used for performing backup and restore
> operations. These gets installed in the image along with some other required
> packages in blueprint.

Reviewer: ```suggestion
### Build a tools image to interact with MongoDB Atlas

We need to build a tools image that should containe `mongosh`, `atlas` and other related
utilities so that it can be used to interact with the MondoDB Atlas database to add some data to it.
This image will also be used in the Blueprint to run the `atlas` backup and restore
related commands against the MongoDB Atlas database. Please execute the commands
below to build and push the image.
```

Author: fixed

```bash

> $ cd ~/kanister/docker/mongodb-atlas
> $ docker build -t <registry>/<account_name>/mongodb-atlas:<tag_name> .
> $ docker push <registry>/<account_name>/mongodb-atlas:<tag_name>

Reviewer: ```suggestion
$ cd ~/kanister/docker/mongodb-atlas
$ docker build -t <registry>/<repository>/mongodb-atlas:<tag_name> .
$ docker push <registry>/<repository>/mongodb-atlas:<tag_name>
```

Author: fixed

```

### Create Blueprint

> Create Blueprint in the same namespace as the controller
> 
> **NOTE:**
> 
> Replace `<registry>`, `<account_name>` and `<tag_name>` for the image with the
> values provided while building MongoDB Atlas docker image in
> ./mongodb-atlas-blueprint.yaml before running following command.

Reviewer: ```suggestion
Create Blueprint in the same namespace as the Kanister controller.

**NOTE:**

Please make sure that the `<registry>`, `<repository>` and `<tag_name>` in the blueprint
(`mongodb-atlas-blueprint.yaml`) are replaced with correct values of the tools image that
we built above.
```

```bash
$ kubectl create -f ./mongodb-atlas-blueprint.yaml -n kanister
```

### Create Secret

Create a Secret which contains Atlas account details.

```bash
$ kubectl create namespace mongodb-atlas-test
namespace/mongodb-atlas-test created

$ kubectl create secret generic mongoatlassecret \
    --from-literal=Public_Key="<public key from Atlas account>" \
    --from-literal=Private_Key="<private key from Atlas account>" \
    --from-literal=Org_Id="<organization ID from Atlas account>" \
    --from-literal=Project_Id="<project ID from Atlas account>" \
    --from-literal=Cluster_Name="<cluster name from Atlas account>" \
    -n mongodb-atlas-test

> secret/mongoatlassecret created

Reviewer: we should also mention are someone, who is following this guide, can figure these values for `Public_key` etc out.

Author: added details about how user can find these fields

```

### Populate data in database

```bash
# Create a collection in database
$ mongosh "mongodb+srv://<cluster name>.<host>/<database name>" --apiVersion 1 \
    --username <username for Atlas account> -p <password for Atlas account> \

>     --quiet --eval "db.people.insertOne({'name': {'first': 'Alan', last: 'Turing'}})"

Reviewer: which database are we connecting to here and who is creating that database?

{
  acknowledged: true,
  insertedId: ObjectId("643812344c2ce812b7aeaccf")
}

# View the people data in the database
$ mongosh "mongodb+srv://<cluster name>.<host>/<database name>" --apiVersion 1 \
    --username <username for Atlas account> -p <password for Atlas account> \
    --quiet --eval "db.people.find()"
[
  {
    _id: ObjectId("643812344c2ce812b7aeaccf"),
    name: { first: 'Alan', last: 'Turing' }
  }
]
```

## Protect the Application

You can now take a backup of the MongoDB data using an ActionSet defining
backup for this application. Create an ActionSet in the same namespace as the
controller.

```bash
$ kanctl create actionset --action backup --namespace kanister \
    --blueprint mongodb-atlas-blueprint \
    --objects v1/secrets/mongodb-atlas-test/mongoatlassecret
actionset backup-tfjps created

# View the status of the actionset
$ kubectl --namespace kanister get actionsets.cr.kanister.io backup-tfjps
NAME           PROGRESS   RUNNING PHASE             LAST TRANSITION TIME   STATE
backup-tfjps   100.00                               2023-04-13T14:40:24Z   complete
```

### Disaster strikes!

> 
> Let's say someone accidentally deleted the people collection:
> ```bash

Reviewer: ```suggestion

Let's say someone accidentally deleted the people collection:

```bash
```

Author: fixed

# MongoDB Atlas

[MongoDB Atlas](https://www.mongodb.com/atlas) is an integrated suite of cloud
database and data services to accelerate and simplify how you build with data.
It deploys and scales a MongoDB cluster in the cloud.

## Prerequisites

* Kubernetes 1.20+
* Kanister controller version 0.91.0 installed in your cluster
* Kanctl CLI installed (https://docs.kanister.io/tooling.html#install-the-tools)
* MongoDB Atlas account registered with organization, project and cluster

## Integrating with Kanister

In case, if you don't have `Kanister` installed already, use following commands
to install.
```bash
$ helm repo add kanister https://charts.kanister.io
$ helm install kanister --namespace kanister --create-namespace \
    kanister/kanister-operator --set image.tag=0.91.0
```

### Build a tools image to interact with MongoDB Atlas

We need to build a tools image that contains `mongosh`, `atlas` and other

> related utilities so that it can be used to interact with the MondoDB Atlas

Author: fixed

database to add data to it.
This image will also be used in the Blueprint to run the `atlas` backup and
restore related commands against the MongoDB Atlas database. Please execute the
commands below to build and push the image.

```bash
$ cd ~/kanister/docker/mongodb-atlas
$ docker build -t <registry>/<repository>/mongodb-atlas:<tag_name> .

> $ docker push <registry>/<repository>/mongodb-atlas:<tag_name>

Author: fixed

```

### Create Blueprint

Create Blueprint in the same namespace as the Kanister controller.

**NOTE:**

Please make sure that the `<registry>`, `<repository>` and `<tag_name>` in the
blueprint (`mongodb-atlas-blueprint.yaml`) are replaced with correct values of
the tools image that we built above.

```bash
$ kubectl create -f ./mongodb-atlas-blueprint.yaml -n kanister
```

### Create Secret

To create a Secret, you need to specify your Atlas account details -

Public_Key: Public key included in Organization level API Key (you'll find it in Organizations > Access Manager > API Keys)

Private_Key: Private key included in Organization level API Key (you'll find it in Organizations > Access Manager > API Keys)

Org_Id: A unique 24 characters Organization ID (you'll find it in Organizations > Settings)

Project_Id: A unique 24 character Project ID (you'll find it in Organizations > Project > Settings)

> Cluster_Name: Cluster name in project (you'll find it in Organizations > Project > Database)

Author: fixed

```bash
$ kubectl create namespace mongodb-atlas-test
namespace/mongodb-atlas-test created

> $ kubectl create secret generic mongoatlassecret \

Author: fixed

    --from-literal=publickey="<Public_Key>" \
    --from-literal=privatekey="<Private_Key>" \
    --from-literal=orgid="<Org_Id>" \
    --from-literal=projectid="<Project_Id>" \
    --from-literal=clustername="<Cluster_Name>" \
    -n mongodb-atlas-test

> secret/mongoatlassecret created

Author: fixed

To insert data into database, you need to provide `connection string`.  You can
find this `connection string` using steps mentioned [here](https://www.mongodb.com/docs/atlas/tutorial/connect-to-your-cluster/#connect-to-your-atlas-cluster)

```bash
# Create a collection in database
$ mongosh "<connection string>" --apiVersion 1 \
    --username <username for Atlas account> -p <password for Atlas account> \
    --quiet --eval "db.people.insertOne({'name': {'first': 'Alan', last: 'Turing'}})"
{
  acknowledged: true,
  insertedId: ObjectId("643812344c2ce812b7aeaccf")
}

# View the people data in the database
$ mongosh "<connection string>" --apiVersion 1 \
    --username <username for Atlas account> -p <password for Atlas account> \
    --quiet --eval "db.people.find()"
[
  {
    _id: ObjectId("643812344c2ce812b7aeaccf"),
    name: { first: 'Alan', last: 'Turing' }
  }
]
```

## Protect the Application

You can now take a backup of the MongoDB data using an ActionSet defining
backup for this application. Create an ActionSet in the same namespace as the
controller.

```bash
$ kanctl create actionset --action backup --namespace kanister \
    --blueprint mongodb-atlas-blueprint \

>     --objects v1/secrets/mongodb-atlas-test/mongoatlassecret

Author: fixed

true

# Try to access this data in the database
$ mongosh "<connection string>" --apiVersion 1 \
    --username <username for Atlas account> -p <password for Atlas account> \
    --quiet --eval "db.people.find()"
# No entries found
```

### Restore the Application

To restore the missing data, use the backup that was created.

```bash
$ kanctl create actionset --action restore --from backup-tfjps --namespace kanister
actionset restore-backup-tfjps-bhv5j created

# View the status of the ActionSet
$ kubectl --namespace kanister get actionsets.cr.kanister.io restore-backup-tfjps-bhv5j
NAME                         PROGRESS   RUNNING PHASE              LAST TRANSITION TIME   STATE
restore-backup-tfjps-bhv5j   100.00                                2023-04-13T15:07:10Z   complete
```

Now the lost data should be visible in the database.

```bash
# Try to access this data in the database
$ mongosh "<connection string>" --apiVersion 1 \
    --username <username for Atlas account> -p <password for Atlas account> \
    --quiet --eval "db.people.find()"
[
  {
    _id: ObjectId("643812344c2ce812b7aeaccf"),
    name: { first: 'Alan', last: 'Turing' }
  }
]
```

### Delete the Artifacts

> The artifacts created by the backup action can be cleaned up using following command.

Author: fixed

# MongoDB Atlas

[MongoDB Atlas](https://www.mongodb.com/atlas) is an integrated suite of cloud
database and data services to accelerate and simplify how you build with data.
It deploys and scales a MongoDB cluster in the cloud.

## Prerequisites

* Kubernetes 1.20+
* Kanister controller version 0.91.0 installed in your cluster
* Kanctl CLI installed (https://docs.kanister.io/tooling.html#install-the-tools)

> * MongoDB Atlas cluster setup (https://www.mongodb.com/docs/atlas/getting-started)

Reviewer: ```suggestion
* Already provisioned MongoDB Atlas cluster (https://www.mongodb.com/docs/atlas/getting-started)
```

Author: fixed

    --quiet --eval "db.people.insertOne({'name': {'first': 'Alan', last: 'Turing'}})"
{
  acknowledged: true,
  insertedId: ObjectId("643812344c2ce812b7aeaccf")
}

# View the people data in the database
$ mongosh "<connection string>" --apiVersion 1 \
    --username <database username> -p <database password> \
    --quiet --eval "db.people.find()"
[
  {
    _id: ObjectId("643812344c2ce812b7aeaccf"),
    name: { first: 'Alan', last: 'Turing' }
  }
]
```

## Protect the Application

You can now take a backup of the MongoDB data using an ActionSet defining
backup for this application. Create an ActionSet in the same namespace as the
controller.

```bash
$ kanctl create actionset --action backup --namespace kanister \
    --blueprint mongodb-atlas-blueprint \
    --objects v1/secrets/mongodb-atlas-test/mongo-atlas-secret
actionset backup-tfjps created

# View the status of the actionset
$ kubectl --namespace kanister get actionsets.cr.kanister.io backup-tfjps
NAME           PROGRESS   RUNNING PHASE             LAST TRANSITION TIME   STATE
backup-tfjps   100.00                               2023-04-13T14:40:24Z   complete
```

### Disaster strikes!

> Let's say someone accidentally deleted the people collection:

Reviewer: ```suggestion
Let's say someone accidentally deleted the `people` collection:
```

Author: fixed

* Kubernetes 1.20+
* Kanister controller version 0.91.0 installed in your cluster
* Kanctl CLI installed (https://docs.kanister.io/tooling.html#install-the-tools)

> * MongoDB Atlas account registered with organization, project and cluster (https://www.mongodb.com/docs/atlas/getting-started)

Reviewer: 1. By `MongoDB Atlas account`, I am assuming you are talking about user account that was created in MongoDB Atlas cloud? 
2. If yes, should if we registered with organization, project and cluster? What do we exactly mean by registering here? What do we do as part of registering?

Author: yes user account in MongoDB Atlas. We assume Organization, Project and Cluster is already created/setup by user before trying out this BP.

Author: this statement is under prerequisite section. isn't it self-explanatory?
